### PR TITLE
Added New Route to fetch User Stocks

### PR DESCRIPTION
**Pull request by vikasosmium**
Date: 09-Dec-2024

Developer Name: [Vikas Singh](https://github.com/vikasosmium)

---

## Issue Ticket Number

- #2219 
## Description

This PR includes adding one route to fetch User Stocks and a deprecation warning for the old route.
### Documentation Updated?

- [x] Yes
- [ ] No
- Doc Link https://github.com/Real-Dev-Squad/website-api-contracts/pull/203

### Under Feature Flag

- [x] Yes
- [ ] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
**Integration Test**

image

**Unit Test**
image

## Test Coverage

**Integration**
image
models
image
routes
image
validators
image
controllers
image

**Unit**
image
image

### CODE REVIEW

File: controllers/stocks.js [hunk 1]
```diff
@@ -58,8 +70,41 @@ const getSelfStocks = async (req, res) => {
...
+ * @param req {Object} - Express request object
+ * @param res {Object} - Express response object
+ */
+const getUserStocks = async (req, res) => {
+  const { id: authenticatedUserId } = req.userData;
+  const userId = req.params.userId;
+
+  try {
+    if (userId !== authenticatedUserId) {
+      return res.status(403).json({ message: "Unauthorized access" });
```
[the remark this answers is not in the document]
Comment (vikasosmium) [reply controllers/stocks.js:85]: can't find boom for success

[the remark this answers is not in the document]
Comment (vikasosmium) [reply controllers/stocks.js]: fixed it!

File: controllers/stocks.js [hunk 2]
```diff
@@ -58,8 +70,41 @@ const getSelfStocks = async (req, res) => {
...
+    }
+
+    const userStocks = await stocks.fetchUserStocks(userId);
+
+    if (userStocks.length === 0) {
+      logger.info(`No stocks found for user ${userId}`);
+      return res.status(204).json({ message: "No stocks found", userStocks: [] });
```
[the remark this answers is not in the document]
Comment (vikasosmium) [reply controllers/stocks.js]: fixed it!

File: controllers/stocks.js [hunk 3]
```diff
@@ -44,10 +44,49 @@ const fetchStocks = async (req, res) => {
...
+ * This API is kept temporarily for backward compatibility.
+ */
 const getSelfStocks = async (req, res) => {
   try {
     const { id: userId } = req.userData;
     const userStocks = await stocks.fetchUserStocks(userId);
+
+    res.set(
+      "X-Deprecation-Warning",
+      "WARNING: This endpoint is being deprecated and will be removed in the future. Please use `/stocks/:userId` route to get the user stocks details."
```
Comment (Mir-SA) [inline controllers/stocks.js:61]: @vikasosmium this res header gets set properly, but for some it always returns empty list of stocks from both routes, old one & new one, even though i've added multiple stocks for multiple users...
is it a bug or intended?

Comment (vikasosmium) [reply controllers/stocks.js:61]: not a bug, it's a feature, it will only allow you to get your data.

**Review by Mir-SA — APPROVED** [review]

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by Achintya-Chatterjee — DISMISSED** [review]

File: test/integration/stocks.test.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+  afterEach(async function () {
+    await cleanDb();
+    sinon.restore();
+  });
+
+  it("Should return user stocks when stocks are available", function (done) {
+    const userStocks = [{id: "5YGjUSW1SinwCNfuLXOO", userId: "DHLG3gYMTtMenj6lciWz", stockId: "s2eYDswDUAoQxwAhh07f", stockName: "EURO", quantity: 1, orderValue: 150, initialStockValue: 150}];
+
+    sinon.stub(stocks, "fetchUserStocks").resolves(userStocks);
```
Comment (prakashchoudhary07) [inline test/integration/stocks.test.ts:33]: Why is this stubed?
We have a db emulator in place, so we should not be mocking it

File: test/integration/stocks.test.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+        expect(res.body.userStocks).to.deep.equal(userStocks);
+
+        return done();
+      });
+  });
+
+  it("Should return empty object when no stocks are found", function (done) {
+    const userStocks=[];
+
+    sinon.stub(stocks, "fetchUserStocks").resolves(userStocks);
```
Comment (prakashchoudhary07) [inline test/integration/stocks.test.ts:55]: RE

**Review by prakashchoudhary07 — CHANGES_REQUESTED** [review]

File: test/integration/stocks.test.ts (hunk 4 above)
Comment (vikasosmium) [reply test/integration/stocks.test.ts]: done

**Review by Achintya-Chatterjee — APPROVED** [review]

**Review by tejaskh3 — APPROVED** [review]

**Comment by tejaskh3** [comment]
Before merging this PR please add test coverage for the test cases you have written recently for this PR.

**Comment by vikasosmium** [comment]
> Before merging this PR please add test coverage for the test cases you have written recently for this PR. Rest looks good to me

done

**Review by prakashchoudhary07 — APPROVED** [review]

**Pull request merged**